src/simulations/collect.jl

```diff
@@ -0,0 +1,285 @@
+## Basic data collection functions
+
+get_data(a, s::Symbol) = getproperty(a, s)
+get_data(a, f::Function) = f(a)
+
+"""
+    collect_agent_data(model::ABM, properties::AbstractArray, step::Int)
+
+Collect agent properties into a dataframe. `properties` can have symbols (agent fields) or functions that take an agent as input.
+"""
+function collect_agent_data(model::ABM, properties::AbstractArray, step::Int)
+  dd = DataFrame()
+  dd[!, :id] = collect(keys(model.agents))
+  for fn in properties
+    dd[!, fn] = get_data.(values(model.agents), fn)
+  end
+  dd[!, :step] = fill(step, size(dd, 1))
+  return dd
+end
+
+"""
+    collect_model_data(model::ABM, properties::AbstractArray, step::Int)
+
+Collect model properties from functions or symbols provided in `properties`.
+"""
+function collect_model_data(model::ABM, properties::AbstractArray, step::Int)
+  dd = DataFrame()
+  for fn in properties
+    r =  get_data(model, fn)
+    dd[!, Symbol(fn)] = [r]
+  end
+  dd[!, :step] = fill(step, size(dd, 1))
+  return dd
+end
+
```

Comment:
there used to be in place functions `collect_.._data!(df, ...)` , what happened to them?

Reply:
They are supposed to be part of the public API in this new design.

Reply:
They were redundant. We have three basic functions now: collect agent data, collect model data, and aggregate. Updating an existing `df` is just as easy as `vcat(df, new_df)`.